### ISSUE TITLE
Unactivated users get wrong error message

### ISSUE THREAD

**Issue by AIC-BV** [wintercms/wn-user-plugin#35]
Hello,

My users have to activate their account before they can sign in. ([Activation type: User](https://github.com/wintercms/wn-user-plugin#activation))
This is to protect guest users from [evil people 'stealing' their accounts](https://github.com/wintercms/wn-user-plugin#guest-users). (my way of handling the important text block)

When a user tries to log in, that didn't activate his account yet, they get an error message like
_The details you entered did not match our records. Please double-check and try again._

This error is wrong, and it should be something like
_You didn't activate your account yet. Please check your email and activate your account._
and, ideally, resend the activation email.

User exists in backend user list, in database and with the [onCheckEmail method](https://github.com/wintercms/wn-user-plugin#checking-if-a-login-name-is-already-taken). So I can change my password, but can never log in because I forgot to activate my account.

The current error message sends me in an infinite loop.
1. It basically tells me to register
2. But when registering it says my account already exists
3. Which suggests me to reset my password
4. Which when you do, you cannot log in because you are not activated yet
5. Which resets the whole loop

**Comment by AIC-BV** [comment]
hmm I am very confused now because it does seem to work locally.
Can be ignored for now ^^

**Comment by AIC-BV** [comment]
> 

Yes, Thanks for pointing that out!
I tested it and it is indeed a 'problem' with `debug mode true/false`.

When it is true, it is throwing the expected answers.
For example:
- ... is suspended
- ... is not yet activated
- ...

But as you mentioned, when debug mode is `false`, it is throwing that generic message.

I understand your argument, but then how would I let my user know their account is not yet activated when debug mode is false? They cannot proceed as long as it is not activated, so it is a really important step. They get a popup to activate their account when registering but some people just ignore the popup and click it away and then 'complain' they cannot log in

**Comment by LukeTowers** [comment]
It depends on your specific use case and users, one idea might be to resend the activation email when they attempt to login. Another (probably better approach) would be to check if the user is activated immediately after successfully validating their credentials but before authenticating their session or sending any auth cookies back and if they aren't activated at that point you can send the more helpful error message without being concerned about the security implications.

**Comment by AIC-BV** [comment]
The second one sounds a bit complex,
but the first option sounds good enough to capture the 1% of users struggling with this issue.

Do I understand you correctly that I have to do some magic to check wether the user is activated or not inside the following event:
```
Event::listen('winter.user.beforeAuthenticate', function($component, $credentials) {
    $login = array_get($credentials, 'login');
    if (!$user = Auth::findUserByLogin($login)) {
        return;
    }

   // check if user is activated or not to throw a proper notifcation with an optional button to resend the activation e-mail if not yet activated
});
```

**Comment by LukeTowers** [comment]
Yes, that would probably work.

**Comment by AIC-BV** [comment]
@LukeTowers
My solution is the following:
```
// extending login - make sure user is activated. If not, resend activation e-mail
Event::listen('winter.user.beforeAuthenticate', function($component, $credentials) {
    $login = array_get($credentials, 'login');
    $user = Auth::findUserByLogin($login);
    if (!$user) return;
        
    // if user is NOT activated, show a popup with a button to resend activation e-mail
    if (!$user->is_activated) {

        // save email to know when to show the popup and feed the onSendActivationEmail button
        Session::put('user.not_activated', $login);

        // refresh page to display message
        // NOT WORKING :(
        return redirect()->refresh();

    }
});
```

```
==
<?php
public function onStart()
{
    $this['user_not_activated'] = Session::pull('user.not_activated', false);
}
?>
==
```
```
{% if user_not_activated %}
    <button data-request='onSendActivationEmail' data-request-data='email: "{{ user_not_activated }}"'>Resend activation email</button>
{% endif %}
```

The code above (except the refresh 😞) would work, BUT;
Guest users that require activation before login are currently not allowed to request an activation e-mail because you have to be logged in: https://github.com/wintercms/wn-user-plugin/blob/main/components/Account.php#L476

To solve this, I added the following code:
```
public function onSendActivationEmail()
{
    try {
        $user = $this->user();
        if (!$user) {
                
            // check for login (fix guest users)
            $login = Input::get('email', false);
            if ($login) $user = Auth::findUserByLogin($login);
            if (!$user) throw new ApplicationException(Lang::get(/*You must be logged in first!*/'winter.user::lang.account.login_first'));
        }
        
        ...
```
Would this, or something like this, be accepted in PR?
It is a must to support this toggle?
![image](https://github.com/wintercms/wn-user-plugin/assets/89913092/2fed1d89-7664-4c51-a477-61424da48ef9)

**Comment by AIC-BV** [comment]
Only today I have 3 new users, that didn't activate their account yet
![image](https://github.com/wintercms/wn-user-plugin/assets/89913092/51bccd94-f3b1-4ecf-bd2b-14cf51cb7278)

**Comment by AIC-BV** [comment]
Only today I have 3 users that didn't activate their account yet
![image](https://github.com/wintercms/wn-user-plugin/assets/89913092/413315a6-b436-4c5a-a6ac-69813d77a5eb)

**Comment by LukeTowers** [comment]
> Isn't that what [onCheckEmail](https://github.com/wintercms/wn-user-plugin#checking-if-a-login-name-is-already-taken) does? 😅 (letting a bad actor to be able to figure out who has an account on your service or not)

@AIC-BV yes but it is not provided in the core plugin for a reason, it's up to the developer to make the decision if that is an acceptable risk or not.

**Comment by bennothommo** [comment]
@AIC-BV I've used the User plugin several times with the User activation enabled, and I cannot say I've had many people ask why they could not log in that turned out to be an activation issue (maybe 3 or 4 in the same amount of years?).

It's all about how you onboard them after completing the registration - you could perhaps redirect them to a content page that welcomes them to the site/app and tells them to check their email for the activation message? Or write a Snowboard plugin that increases the timeout of the flash message?

**Comment by AIC-BV** [comment]
> Member

After registration there is a popup saying they have to activate their account via email before they can log in
Some people just ignore/don't pay attention to it...
